### PR TITLE
Add field 'co' in databases for OpenTSDB 

### PR DESCRIPTION
**Pull request by yac01**
DO NOT DELETE THIS TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`

### CODE REVIEW

File: includes/opentsdb.inc.php [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        global $config, $opentsdb;      // Gloabal Variable look config.php for enable or disable OpenTSDB
+
+       if ($config['opentsdb']['enable'] == true) {   // if OpenTSDB enable
+             if ($opentsdb != true) {  // if the connection is already made
+                $opentsdb = fsockopen($config['opentsdb']['host'], $config['opentsdb']['port']);
+              }  
+        $timestamp = time();
+        $tmp_tags = "hostname=".$device['hostname'];
+        
+        /*If you want to sort your metrics by customers and gain speed when you send yous queries on grafana you need to
```
Comment (laf) [inline includes/opentsdb.inc.php]: You can remove this comment.

File: includes/opentsdb.inc.php [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+ */
+ 
+ 
+ function opentsdb_update($device, $measurement, $tags, $fields) {
+
+        global $config, $opentsdb;      // Gloabal Variable look config.php for enable or disable OpenTSDB
+
+       if ($config['opentsdb']['enable'] == true) {   // if OpenTSDB enable
+             if ($opentsdb != true) {  // if the connection is already made
+                $opentsdb = fsockopen($config['opentsdb']['host'], $config['opentsdb']['port']);
```
Comment (laf) [inline includes/opentsdb.inc.php]: Can you add a `d_echo()` here to say that this is connecting / connected for debug purposes.

File: includes/opentsdb.inc.php [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+ * option) any later version.  Please see LICENSE.txt at the top level of
+ * the source code distribution for details.
+ */
+ 
+ 
+ function opentsdb_update($device, $measurement, $tags, $fields) {
+
+        global $config, $opentsdb;      // Gloabal Variable look config.php for enable or disable OpenTSDB
+
+       if ($config['opentsdb']['enable'] == true) {   // if OpenTSDB enable
```
Comment (laf) [inline includes/opentsdb.inc.php]: Can you add a `d_echo()` here to say opentsdb is enabled.

File: includes/polling/ports.inc.php [hunk 4]
```diff
@@ -747,6 +747,10 @@
 
         influx_update($device, 'ports', rrd_array_filter($tags), $fields);
         graphite_update($device, 'ports|' . $ifName, $tags, $fields);
+           foreach ($fields as $k => $v) {
+                opentsdb_update($device, "port." . $k, array('ifName' => $this_port['ifName'],'ifIndex' =>  getPortRrdName($port_id)), array('key' => $v));
```
Comment (laf) [inline includes/polling/ports.inc.php]: Any reason this is here and not just done in `opentsdb_update()` and each metric has to be sent separately?

File: config.php.default [hunk 5]
```diff
@@ -8,6 +8,13 @@ $config['db_user'] = 'USERNAME';
 $config['db_pass'] = 'PASSWORD';
 $config['db_name'] = 'librenms';
 
+
+// OpenTSDB configuration
```
Comment (laf) [inline config.php.default]: You should move all of this into includes/defaults.inc.php but you must set `$config['opentsdb']['enable'] = false;` as people won't have this setup to use by default.

File: includes/polling/ports.inc.php [hunk 6]
```diff
@@ -744,10 +744,14 @@
         $fields['ifOutErrors_rate'] = $port['ifOutErrors_rate'];
         $fields['ifInOctets_rate'] = $port['ifInOctets_rate'];
         $fields['ifOutOctets_rate'] = $port['ifOutOctets_rate'];
+        $fields['ifInBits'] = $port['stats']['ifInBits_rate'];
```
Comment (murrant) [inline includes/polling/ports.inc.php]: Why are these needed?  If you want bits, you can just take ifOutOctets_rate / 8 when graphing the data.

Comment (romain001) [inline includes/polling/ports.inc.php]: how do you do this with grafana ??

**Comment by murrant** [comment]
@yac01 You cannot modify build.sql, it is in the past.

Create a new file in sql-schema, `200.sql` for example.  And put your alter commands in there. 
Run it to alter your database.
Run `./scripts/build-schema.php`  (running `./scripts/pre-commit.php` will tell you to do this)

File: includes/polling/ports.inc.php (hunk 6 above)
Comment (murrant) [inline includes/polling/ports.inc.php]: Just use `math (*8)` in the SELECT

**Comment by laf** [comment]
Also @yac01, can you please keep discussion on this PR here please - opening more threads and PMs on the community site doesn't help :(

File: includes/polling/ports.inc.php (hunk 6 above)
Comment (yac01) [inline includes/polling/ports.inc.php]: you can't do this with opentsdb I have already searched with InfluxDB it's right you can do this  
op2

gm

Comment (laf) [inline includes/polling/ports.inc.php]: Are you 100% sure this isn't possible? We'd prefer not to send calculated metrics across as this is something that the front end should deal with.

File: sql-schema/200.sql [hunk 7]
```diff
@@ -0,0 +1 @@
+ALTER TABLE `devices` ADD `co` varchar(80) CHARACTER SET latin1 NULL DEFAULT '0';
```
Comment (laf) [inline sql-schema/200.sql]: You need to allow people to edit this within the Webui as well.

File: doc/Extensions/OpenTSDB.md [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+### Requirements
+ - OpenTSDB
+ - Grafana
+ 
+### What you don't get
+ - Pretty graphs, this is why at present you need Grafana. You need to build your own graphs within Grafana.
+ 
+RRD will continue to function normally so LibreNMS itself should continue to function normally.
+
+configuration in /opt/librenms/includes/defaults.inc.php.
```
Comment (laf) [inline doc/Extensions/OpenTSDB.md]: Configuration should be done in config.php so point to that rather than defaults.inc.php

File: build.sql [hunk 9]
```diff
@@ -10,7 +10,7 @@ CREATE TABLE IF NOT EXISTS `bill_ports` ( `bill_id` int(11) NOT NULL, `port_id`
...
...` varchar(16) COLLATE utf8_unicode_ci NOT NULL DEFAULT 'udp', `timeout` int(11) DEFAULT NULL, `retries` int(11) DEFAULT NULL, `bgpLocalAs` varchar(16) CHARACTER SET latin1 DEFAULT NULL, `sysDescr` text CHARACTER SET latin1, `sysContact` text CHARACTER SET latin1, `version` text CHARACTER SET latin1, `hardware` text CHARACTER SET latin1, `features` text CHARACTER SET latin1, `location` text COLLATE utf8_unicode_ci, `os` varchar(32) COLLATE utf8_unicode_ci DEFAULT NULL, `status` tinyint(4) NOT NULL DEFAULT '0', `ignore` tinyint(4) NOT NULL DEFAULT '0', `disabled` tinyint(1) NOT NULL DEFAULT '0', `uptime` bigint(20) DEFAULT NULL, `agent_uptime` int(11) NOT NULL, `last_polled` timestamp NULL DEFAULT NULL, `last_polled_timetaken` double(5,2) DEFAULT NULL, `last_discovered_timetaken` double(5,2) DEFAULT NULL, `last_discovered` timestamp NULL DEFAULT NULL, `purpose` varchar(64) COLLATE utf8_unicode_ci DEFAULT NULL, `type` varchar(20) COLLATE utf8_unicode_ci NOT NULL, `serial` text COLLATE utf8_unicode_ci, PRIMARY KEY (`device_id`), KEY `status` (`status`), KEY `hostname` (`hostname`), KEY `sysName` (`sysName`), KEY `os` (`os`) ) ENGINE=InnoDB  DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci;
```
Comment (laf) [inline build.sql]: You've still got your change here, please revert this.

File: misc/db_schema.yaml [hunk 10]
```diff
@@ -426,7 +427,7 @@ devices:
     sysContact: { Field: sysContact, Type: text, 'Null': true, Default: null, Extra: '' }
     sysDescr: { Field: sysDescr, Type: text, 'Null': true, Default: null, Extra: '' }
     sysName: { Field: sysName, Type: varchar(128), 'Null': true, Default: null, Extra: '' }
-    sysObjectID: { Field: sysObjectID, Type: varchar(128), 'Null': true, Default: null, Extra: '' }
+    sysObjectID: { Field: sysObjectID, Type: varchar(64), 'Null': true, Default: null, Extra: '' }
```
Comment (laf) [inline misc/db_schema.yaml]: It looks like your database schema is not up to date, please update, run validate.php and fix the issues present then rebuild db_schema.yaml

File: doc/Extensions/OpenTSDB.md (hunk 8 above)
Comment (yac01) [inline doc/Extensions/OpenTSDB.md]: You have already said to put it the configuration in defaults.inc.php the last time?? when i will put it config or defaults ??? or the two ? 
put

File: includes/polling/ports.inc.php (hunk 6 above)
Comment (yac01) [inline includes/polling/ports.inc.php]: yes i had posted this in grafana community   and opentsdb don't support that https://community.grafana.com/t/how-can-i-use-math-8-with-opentsdb-like-influxdb/1770

File: doc/Extensions/OpenTSDB.md (hunk 8 above)
Comment (yac01) [inline doc/Extensions/OpenTSDB.md]: @laf  
Do I have to put them in both default and config file or just in one of 2 or default I have to put it in default ????

Comment (laf) [inline doc/Extensions/OpenTSDB.md]: No, this is the docs. You're telling people config is in includes/defaults.inc.php, users should never be directed to this file and all changes should be in config.php

**Comment by laf** [comment]
@yac01 Can you pick one of your PRs to stick with and close the other.

File: sql-schema/200.sql (hunk 7 above)
Comment (yac01) [inline sql-schema/200.sql]: I didn't know that we could modify the databases  with the webui ?? great that but how we do this ?
